### PR TITLE
monkey patch logging to disable tracebacks if log.traceback_level == debug

### PR DESCRIPTION
**Pull request by yanchenko-igor**
#1265 I came to solution how to disable tracebacks by managing if from config. What do you think about it?

### CODE REVIEW

File: patroni/log.py [hunk 1]
```diff
@@ -61,6 +67,7 @@ def emit(self, record):
 class PatroniLogger(Thread):
 
     DEFAULT_LEVEL = 'INFO'
+    TRACEBACK_LEVEL = 'ERROR'
```
Comment (CyberDem0n) [inline patroni/log.py]: `DEFAULT_TRACEBACK_LEVEL`

File: patroni/log.py [hunk 2]
```diff
@@ -11,6 +11,12 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+def debug_exception(logger_obj, msg, *args, **kwargs):
+    kwargs.pop("exc_info", False)
+    logger_obj.error(msg, *args, exc_info=False, **kwargs)
+    logger_obj.debug(msg, *args, exc_info=True, **kwargs)
```
Comment (CyberDem0n) [inline patroni/log.py]: Basically that means we will see tracebacks only if log level is set to debug, which is too verbose.

File: patroni/log.py [hunk 3]
```diff
@@ -99,6 +106,8 @@ def reload_config(self, config):
                 self._queue_handler.queue.maxsize = config.get('max_queue_size', self.DEFAULT_MAX_QUEUE_SIZE)
 
             self._root_logger.setLevel(config.get('level', PatroniLogger.DEFAULT_LEVEL))
+            if config.get('traceback_level', PatroniLogger.TRACEBACK_LEVEL).lower() == 'debug':
```
Comment (CyberDem0n) [inline patroni/log.py]: The `traceback_level` could be changed in the runtime.

File: patroni/log.py (hunk 2 above)
Comment (yanchenko-igor) [reply patroni/log.py]: Only if we set log.traceback_level == debug, if we don't change log.traceback_level nothing changes.

File: patroni/log.py (hunk 1 above)
Comment (yanchenko-igor) [reply patroni/log.py]: fixed

File: patroni/log.py (hunk 3 above)
Comment (yanchenko-igor) [reply patroni/log.py]: made it switch back to original behavior if configuration is changed back.

File: patroni/log.py [hunk 4]
```diff
@@ -99,6 +111,10 @@ def reload_config(self, config):
                 self._queue_handler.queue.maxsize = config.get('max_queue_size', self.DEFAULT_MAX_QUEUE_SIZE)
 
             self._root_logger.setLevel(config.get('level', PatroniLogger.DEFAULT_LEVEL))
+            if config.get('traceback_level', PatroniLogger.DEFAULT_TRACEBACK_LEVEL).lower() == 'debug':
+                logging.Logger.exception = debug_exception
+            elif config.get('traceback_level', PatroniLogger.DEFAULT_TRACEBACK_LEVEL).lower() == 'error':
```
Comment (CyberDem0n) [inline patroni/log.py]: I think here it should be just `else`.

**Comment by CyberDem0n** [comment]
It would be also good to:
1. add [support](https://github.com/zalando/patroni/blob/master/patroni/config.py#L245-L246) of `PATRONI_LOG_TRACEBACK_LEVEL`
2. The documentation
3. Tests

File: docs/SETTINGS.rst [hunk 5]
```diff
@@ -46,6 +46,7 @@ Global/Universal
 Log
 ---
 -  **level**: sets the general logging level. Default value is **INFO** (see `the docs for Python logging <https://docs.python.org/3.6/library/logging.html#levels>`_)
+-  **traceback_level**: sets the level where tracebacks will be visible. Default value is **ERROR**. Set it to **DEBUG** if you want to see tracebacks only if you enable **log.level=DEBUG**.
```
Comment (CyberDem0n) [inline docs/SETTINGS.rst]: We need the same here: https://github.com/zalando/patroni/blob/master/docs/ENVIRONMENT.rst#log

**Comment by FxKu** [comment]
:+1:

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**